### PR TITLE
Bump minimum version to go 1.21.0

### PR DESCRIPTION
**Pull request by liustanley**
**Description:** <Describe what has changed.>

We currently use `go 1.21` in all go.mod files, this PR changes all go.mod files to include the minor version by using `go 1.21.0`. It seems that using the minor version is recommended by the Go project: https://github.com/golang/go/issues/62278. One of the dependencies in collector-contrib also uses `go 1.21.0`, so this will need to be updated eventually anyways: https://github.com/cilium/ebpf/blob/main/go.mod#L3.

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** <Describe what testing was performed and which tests were added.>

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: .github/workflows/build-and-test.yml [hunk 1]
```diff
@@ -291,13 +291,13 @@ jobs:
...
-        if: startsWith( matrix.go-version, '1.21' ) != true
+        if: startsWith( matrix.go-version, '1.21.0' ) != true
         run: make gotest GROUP=${{ matrix.group }}
       - name: Run Unit Tests With Coverage
-        if: startsWith( matrix.go-version, '1.21' ) # only run coverage on one version
+        if: startsWith( matrix.go-version, '1.21.0' ) # only run coverage on one version
         run: make gotest-with-cover GROUP=${{ matrix.group }}
       - uses: actions/upload-artifact@v4
-        if: startsWith( matrix.go-version, '1.21' ) # only upload artifact for one version
+        if: startsWith( matrix.go-version, '1.21.0' ) # only upload artifact for one version
```
Comment (atoulme) [inline .github/workflows/build-and-test.yml:300]: this doesn't seem related to your change

File: .chloggen/go-1.21.0.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: breaking
```
Comment (atoulme) [inline .chloggen/go-1.21.0.yaml:4]: Please clarify how this is a breaking change. Is this a breaking change for the api changelog only?

**Comment by liustanley** [comment]
> Is there a rationale for this change?

The motivation is that we want to import datadog-agent modules to use in collector-contrib, but both repositories have different go versions in their go.mod files. We are making a [change in datadog-agent](https://github.com/DataDog/datadog-agent/pull/24688) to modify the version to match collector-contrib's version, but datadog-agent requires the minor version due to the EBPF dependency. collector-contrib also has this dependency, and the Go project seems to prefer using the minor version as well.

File: .chloggen/go-1.21.0.yaml (hunk 2 above)
Comment (liustanley) [reply .chloggen/go-1.21.0.yaml:4]: This is a breaking change for anyone importing collector-contrib modules, since the minimum version is now required in `go.mod`.

**Comment by crobert-1** [comment]
Related: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32356

**Review by atoulme — APPROVED** [review]
LGTM

**Comment by atoulme** [comment]
CI fails with https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8709569766/job/23889804537?pr=32451 please take a look

**Comment by mx-psi** [comment]
For further context: the Datadog Agent has `go 1.21.0` because it depends on cilium/ebpf that uses `go 1.21.0` https://github.com/cilium/ebpf/blob/32b95e4a83690a69eb2de5796db313893f71d1d0/go.mod#L3

**Review by songy23 — APPROVED** [review]

**Comment by liustanley** [comment]
> CI fails with https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/8709569766/job/23889804537?pr=32451 please take a look

I ran `make genotelcontribcol` with the same version as the CI (go 1.21.9), and it added `toolchain go1.21.9` in `cmd/otelcontribcol/go.mod`. @mx-psi was able to reproduce this on his local machine as well, is this something we need to fix in the CI or can we make this change?

**Review by mx-psi — APPROVED** [review]
I have tried to revert this change upstream on cilium/ebpf/discussions/1438 but it doesn't seem like this is going to change soon.

While this is not well documented, and it is a feature for applications and not libraries, I think this is okay to merge because:
-  The only impact it has on downstream consumers is that we drop support for prereleases of Go 1.21 (so e.g. go1.21rc1 will refuse to build a module with this version). This is mentioned in the 'Go versions' section of https://go.dev/doc/toolchain. This seems okay to me.
- We will probably end up with this change one way or another, since (i) the 'stable' solution in the ecosystem is for everyone to specify the bugfix version and (ii) `go mod init` now adds the bugfix version for you on recent versions.

If the upstream discussion changes, we can revert this, or even do so on the bump to Go 1.22 when Go 1.23 is released
